### ISSUE TITLE
Enhance reliability of feature synchronization

### ISSUE THREAD

**Issue by aaemnnosttv** [google/site-kit-wp#7531]
## Feature Description

Features that are controlled by remote activation are currently only refreshed via a WP cron job twice daily after being received as part of the initial setup.

For most sites this works well, although some sites may have WP cron disabled without properly configuring WP cron to run by an alternative method which can leave a site without important features until the flag is removed entirely, which can be an unknown length of time after the feature is made generally available.

Site Kit should have an alternative method to synchronize its enabled features that does not rely on WP cron.

This issue is related to #6015 but independent from a requirements perspective.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* The plugin should provide a fallback mechanism for synchronizing remote features that does not rely on or invoke WP cron
	* The fallback should ensure features are synchronized if the last sync happened more than 24 hours ago
	* The fallback sync should not happen as part of a page load request

## Implementation Brief

Some context: Since remote features are updated only after cron sync happens, we can't listen to the `on_change` to check if update happend or not. So this approach handles the timestamp check on `admin_init` and fires a non-blocking request that in turn invokes the CRON callback action.

* [ ] Add `includes/Core/Remote_Features/REST_Remote_Features_Controller.php`
  * Define a new route, say `POST:sync-remote-features`  and as a callback invoke cron action `do_action( Remote_Features_Sync::CRON_ACTION )`
* [ ] Update `Google\Site_Kit\Core\Remote_Features\Remote_Features`
  * Update `get_default` method
    * Expand empty array to the two keys: `features` - holds an empty array, and `featuresLastSyncedAt` - `0` by default
  * Update `get_sanitize_callback` method, to take new keys into the account, and add check for `featuresLastSyncedAt` if it is not integer, set value back to `0`
* [ ] In `Google\Site_Kit\Core\Remote_Features\Remote_Features_Activation`
  * Update `get_features`, and `filter_features` to expect features array in `$features['features']` instead of `$features`
* [ ] In `Google\Site_Kit\Core\Remote_Features\Remote_Features_Provider`
  * Update `register` method
    *  Instantiate `REST_Remote_Features_Controller` class and invoke it's `register` method
  * Update `on_admin_init` method
    * Add a check if `featuresLastSyncedAt` timestamp is older than 24h to invoke `wp_remote_post` for a new REST route and pass `blocking => false` as an argument, so it is a non-blocking request.
* [ ] In `Google\Site_Kit\Core\Remote_Features\Remote_Features_Sync`
  * Update `pull_remote_features` method, to set the `features` option correctly to the `features` key instead of the main option as now. And update `featuresLastSyncedAt` option with the current timestamp

### Test Coverage

* Add basic coverage for the REST controller
* Update tests for `Remote_Features_Provider` to include new additions around option structure

### QA:Eng
* Setup Site Kit
* Empty the content of `googlesitekitpersistent_remote_features` DB column from `wp_options` table
* Reload the page, wait for 2 minutes (so first WP core `heartbeat` AJAX request is sent), verify that `googlesitekitpersistent_remote_features` is populated again. 
* Verify that fallback is not executed if `last_synced_at` timestamp is less than 24h ago. This is example of how data would look like:
```
a:20:{s:18:"adBlockerDetection";a:1:{s:7:"enabled";b:1;}s:9:"adsModule";a:1:{s:7:"enabled";b:1;}s:14:"adsenseSetupV2";a:1:{s:7:"enabled";b:1;}s:15:"conversionInfra";a:1:{s:7:"enabled";b:1;}s:16:"dashboardSharing";a:1:{s:7:"enabled";b:1;}s:19:"enhancedMeasurement";a:1:{s:7:"enabled";b:1;}s:19:"ga4ActivationBanner";a:1:{s:7:"enabled";b:0;}s:12:"ga4Reporting";a:1:{s:7:"enabled";b:1;}s:8:"ga4setup";a:1:{s:7:"enabled";b:1;}s:10:"gteSupport";a:1:{s:7:"enabled";b:1;}s:14:"helpVisibility";a:1:{s:7:"enabled";b:1;}s:13:"ideaHubModule";a:1:{s:7:"enabled";b:0;}s:10:"keyMetrics";a:1:{s:7:"enabled";b:1;}s:14:"serviceSetupV2";a:1:{s:7:"enabled";b:1;}s:16:"unifiedDashboard";a:1:{s:7:"enabled";b:1;}s:12:"userFeedback";a:1:{s:7:"enabled";b:1;}s:17:"widgets.dashboard";a:1:{s:7:"enabled";b:1;}s:21:"widgets.pageDashboard";a:1:{s:7:"enabled";b:1;}s:14:"zeroDataStates";a:1:{s:7:"enabled";b:1;}s:15:"last_updated_at";i:1723720515;}
```
At the last item you will notice this value `"last_updated_at";i:1723720515;`. You can modify timestamp (`1723720515` in this case) manually to verify that fallback is called and timestamp updated when value is over 24h, using https://timestampgenerator.com/ or similar websites. Makes sure that only number value is replaced, without space at the end as it will make value invalid. To verify that callback was not invoked, you will see that your updated timestamp value is still present after re-visiting dashboard and waiting for next `heartbeat` sync (approximately 1-2min wait on the Site Kit dashboard)

## Changelog entry

* Add fallback for remote feature activation.

**Comment by techanvil** [comment]
Hey @jimmymadon, I think it would be worth revising and fleshing out the AC here a bit.

- For one thing, taking a step back from the specific requirement relating to feature synchronisation, we also use the WP cron to [synchronise the GA4 property](https://github.com/google/site-kit-wp/blob/0b4219692d02963755cf6cd4b604837f303c7715/includes/Modules/Analytics_4/Synchronize_Property.php#L101-L105). It's entirely likely we'll want to add more cron jobs in future too. We might as well take the opportunity to add our own general purpose cron fallback that works for the GA4 property sync too and future proofs us for requirements down the line.
- We should be explicit about the conditions under which we want the cron alternative to run (i.e. when WP cron is disabled), and to be clear about how often it should run (it should presumably be run to the same schedule as it is now, but we should make that clear in the AC).
- Seeing as the WP cron can be implemented by other means, we should also ensure that our cron alternative can be disabled by the user in case we can't detect that the cron jobs are still running.

**Comment by techanvil** [comment]
Thanks @jimmymadon.

> @techanvil All good points. I think the period would have to be amendable per task? This might complicate the issue slightly since some of the potential IBs for the Remote Features sync could simply rely on certain actions/events in the plugin rather than a specific heartbeat period. I feel we will essentially be creating a dismissed-notification style system that relies on running certain tasks if a certain time has already passed. I'm not against this and this would be quite helpful as it will be generic.

I think the most straightforward approach to take would be to provide our own API modelled on the WP `schedule` function calls that we use. So essentially our own version of `wp_next_scheduled()`, `wp_schedule_event()`, `wp_schedule_single_event()` and `wp_clear_scheduled_hook()`, each of which calls its corresponding function if WP cron is enabled, otherwise calls our own implementation (unless disabled). That should avoid any conceptual ambiguity, and we can flesh out the underlying scheduling mechanism in the IB (unless you prefer to get into it a bit in the AC, which would be fair too).

> UPDATE: Would we need some UI elements to allow the user to disable the running of these tasks? If yes, then we'd need UX's input and this feature would now be something that the user will have to be made aware of which is not ideal for 'background running tasks'.

When I mean it should be disabled by the user, I meant via a WP filter rather than any UI-side control, I don't think that would be necessary.

**Comment by techanvil** [comment]
Thanks @jimmymadon, that SGTM. AC :white_check_mark:

**Comment by techanvil** [comment]
Hi @zutigrm, thanks for drafting this IB. First off, I would say I think it's a pretty ingenious idea, and I think could be a fitting solution in a different situation. That said, I don't think it's such a good fit for the plugin for the following reasons.

- I think it's too hard to follow for our codebase with the logic interleaved between the copied code and WP core. I can see this being troublesome to debug and maintain.
- There's also an element of risk having it so tightly coupled with the current WP cron implementation, which is subject to change outside of our control.
- There's too much chance of affecting the real WP cron. E.g. interfering with its `doing_cron` transient (OK, we could rewrite `spawn_cron()` too, but it starts getting to the point where we're copying half of WP cron).
- It's too closely linked to WP cron, to the point where we're basically ignoring the user's request to `DISABLE_WP_CRON` and running it anyway, yes, with a reduced schedule, but still the same infrastructure which they've explicitly disabled - seeing requests to `wp-cron.php` in the access log might have a sysadmin tearing their hair out for a bit if they thought they'd disabled it.

You make a good point about potentially needing to duplicate `wp-cron.php`, however, on balance I think it would be preferable to take an approach where we do provide our own infrastructure rather than piggybacking on WP cron for the reasons outlined above. Ideally we could provide a lightweight WP cron-a-like that only implements the bits that are needed by Site Kit.

Of course, we could always consider dialling the requirements back to only cover syncing the enabled features, as per the Feature Description. Seeing as you've raised the point about the extent of the infrastructure work that would be required, let's check in with @aaemnnosttv to see what he thinks before carrying on with this one as currently specced.

**Comment by aaemnnosttv** [comment]
@techanvil First, I think it's important to clarify the behavior of `DISABLE_WP_CRON` as it's a bit of a misnomer; the constant does not disable WP cron entirely – only the mechanism which spawns requests for invoking/running it. By default, if not disabled, WP will make a request to `/wp-cron.php` (see [`spawn_cron`](https://developer.wordpress.org/reference/functions/spawn_cron/)). However, even if set cron can still be run by requesting `/wp-cron.php` by another means, or via wp-cli (`wp cron event run --due-now`) usually powered by a real cron job. This can be preferable to make cron more reliable for time-sensitive tasks which would otherwise depend on web traffic to push things along.

All that to say, I don't think we should be overly concerned with the idea that this mechanism should be conditional based on this constant. Instead, I think we want to focus more on the schedule which should be complementary to WP cron but only for our jobs. This also gives us an opportunity to think about jobs that we may be using WP cron for that we don't really need to run all the time and could be "lazier" – that is, only running when SK/wp-admin is being used.

So I definitely agree that we want to avoid reimplementing or copying substantial parts from WP cron internals but I don't think we need to be entirely decoupled from it either. We could potentially use WP cron for everything still, but as mentioned before, we could consider introducing the concept of a lazy job too.

For tasks that do leverage wp cron (such as feature syncing), we could check when the last time it was run, and if the schedule is due, we can dispatch it. It's probably worth looking at how this is done via wp-cli for individual jobs (`wp cron event run <hook>...`). It could be worth looking at other plugins like WP Crontrol for inspiration as well.

---

I hope that helps. One more thing that comes to mind that we'll want to account for is the current user since there is no user set for normal cron jobs, but if we're invoking this from the client it will have that set by default, so we'll want to make sure our pseudo-cron is as close to the real thing as possible if we're planning on running the same jobs in each context.

**Comment by techanvil** [comment]
Hey @aaemnnosttv, thanks for the detail here. One thing though, although we can presumably still trigger `wp-cron.php` when `DISABLE_WP_CRON` is defined, it does seem that conceptually we're not "supposed" to, as per this comment in the source for [`wp-cron.php`](https://github.com/WordPress/wordpress-develop/blob/b315d4ec015572f17848cad3960b20b7cd4da0f1/src/wp-cron.php#L10-L11):

> * Defining DISABLE_WP_CRON as true and calling this file directly are
> * mutually exclusive and the latter does not rely on the former to work.

Although, admittedly the clarifying comment "and the latter does not rely on the former to work" makes me wonder if the term "mutually exclusive" is actually being applied correctly here... :thinking:. WDYT?

**Comment by aaemnnosttv** [comment]
@techanvil I saw that comment as well and my thought is that if they didn't want it to run if `DISABLE_WP_CRON` was defined as `true`, they could exit early but that constant is never checked in that file so the value appears to be irrelevant once we get to that point. I agree it's a bit confusing but I think we can proceed based on how we can see it works and ignore that comment for now.

**Comment by techanvil** [comment]
Thanks @aaemnnosttv, sounds reasonable. I guess this one's back over to you @zutigrm for a fresh look, taking Evan and my feedback above into account. I know there are still a number of avenues this could go down so please feel free to drop me a line if you want to discuss it further :)

**Comment by zutigrm** [comment]
@aaemnnosttv @techanvil Thank you for the input. I have re-wrote the approach to build our own infrastructure for Scheduling events. I was trying to make it simple and efficient, back to you @techanvil for review

**Comment by zutigrm** [comment]
@techanvil When this makes to EB, please assign it to me :D Thanks

**Comment by techanvil** [comment]
Thanks @zutigrm, the service you have outlined looks pretty good.

That said, I have to confess that the further we get into this the more I'm wondering if we are doing the right thing here.

Based on our conversation to date it looks like actually firing off Cron tasks directly would involve too much copying of the WP Cron internals, hence the design for our own scheduling system which is independent of WP Cron.

What the IB is missing is coverage of the last two AC points:

> - Tasks should run only if WP Cron is disabled.
> - There should be a way to disable the running of these tasks via a WP filter.

I realise though that @aaemnnosttv has said the following which does indicate a steer _away_ from the point about only running if WP Cron is disabled:

> All that to say, I don't think we should be overly concerned with the idea that this mechanism should be conditional based on this constant. Instead, I think we want to focus more on the schedule which should be complementary to WP cron but only for our jobs.

The thing is, WP Cron is the idiomatic way for scheduling tasks in WordPress. It's a core feature, and there is an ecosystem supporting it e.g. plugins where users can inspect their Cron schedule and so forth. It's something we _should_ be able to rely on in the majority of cases. Also, I would imagine that some users or their site admins will have gone to some lengths to get it running on their system to allow the self-routed request mechanism to work, that might not work off the bat for our own implementation.

All that is to say, I do think we should stick with WP Cron as the main scheduler, with the implication being that we _should_ make our scheduler a fallback that is dependent on `DISABLE_WP_CRON`. Assuming we are continuing with the direction of implementing our own scheduler.

So, before we get further into the IBR, I would like to make sure that we're fully aligned on the requirements here and the AC is accurate.

As I mentioned at the top of the comment, having got further into it I do wonder if we're taking the right approach here. Do we have an idea of how many users might actually be affected by this? If we're basically on the margins of our user base it might not be worth spending this much development effort on it, and we might be better off revisiting the core requirement of improving the feature synchronisation reliability as a standalone task.

I would be interested to know what Evan thinks on the above. @aaemnnosttv, please can you let us know your thoughts here?

**Comment by aaemnnosttv** [comment]
> That said, I have to confess that the further we get into this the more I'm wondering if we are doing the right thing here.
>
> Based on our conversation to date it looks like actually firing off Cron tasks directly would involve too much copying of the WP Cron internals, hence the design for our own scheduling system which is independent of WP Cron.

Thanks @techanvil – I agree, I think we've somewhat lost sight of the goal here in pursuit of a more ambitious solution.

I think we can simplify things by decoupling this from WP cron entirely. It shouldn't be concerned with whether or not WP cron is disabled. At the end of the day, a cron is essentially a [`do_action( $hook, ...$args )`](https://github.com/WordPress/WordPress/blob/a5efdd1cc9087ef8cea7cdcb803e70679ba7d1f5/wp-cron.php#L182-L191) and this is about adding a reasonable fallback mechanism rather than mirroring the behavior and timing of WP cron. While it would be nice to enhance this for all crons we use, this has proven to be a bit more complicated and stalled us out here.

Let's keep the focus here on feature synchronization only for now. We essentially just need to trigger the same hook that is registered with cron if a certain amount of time has elapsed since the features were last synchronized (e.g. 24 hours) which can happen by listening for updates to the remote features option, for example. This makes it less different than other synchronization tasks we have and should help to make for a clearer path forward. If wp cron is functioning, this fallback would never run so long as it is set to run later than the usual schedule, and that's no problem.

On a related note, I would like to move #8341 forward first as it makes changes to where features are handled before we proceed with this addition.

With that said, let's update the AC and restart with a new IB (sorry @zutigrm – maybe we can save some of this for later?)

cc: @binnieshah

**Comment by zutigrm** [comment]
Thanks @aaemnnosttv 

> which can happen by listening for updates to the remote features option, for example.

Since option is updated only if sync is successful, and will not be invoked if options don't update, I proposed alternative approach of running the checks and making non blocking request to invoke cron action. 

> With that said, let's update the AC and restart with a new IB (sorry @zutigrm – I do like the abstraction around Events – maybe we can save some of this in https://github.com/google/site-kit-wp/issues/6992?)

I will take a look at it once it lands in IB. Thanks

**Comment by eugene-manuilov** [comment]
Thanks, @zutigrm. IB ✔️

**Comment by zutigrm** [comment]
Parked until dependency #8341 is merged

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Tested on dev environment.
- Tested as per steps mentioned under QAB.
- Verified that the fallback ensures features are synchronized if the last sync occurred more than 24 hours ago.
- Verified The fallback sync  not happen as part of a page load request.
- Verified that the fallback ensures features are not synchronized if the last sync occurred less than 24 hours ago.

<details>

Verified that the fallback ensures features are synchronized if the last sync occurred more than 24 hours ago.

https://github.com/user-attachments/assets/4efd1dde-7b2d-4249-ad08-519ca8322b23

Verified that the fallback ensures features are not synchronized if the last sync occurred less than 24 hours ago.

https://github.com/user-attachments/assets/6884c7fb-33c7-4995-9660-5dfff7ea62c1


</details>
